### PR TITLE
Fixed compatibility with python azure module 5.0.0

### PR DESCRIPTION
**Pull request by Nerigal**
Fixed Compatibility with python azure module 5.0.0
cloud azurearm
utils azurearm
Added cloud_provider parameter to module 
azurearm_computer
azurearm_resource
azurearm_network
azurearm_dns

Users can now use the param cloud_provider to pass all the information from a cloud provider profile 
so they dont have to duplicate informations in the pillar 

- [ ] Docs
- [ ] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

### CODE REVIEW

**Comment by Nerigal** [comment]
@waynew 
windows test are failing and doesn't look like its related to the changes ive made

**Comment by waynew** [comment]
@Nerigal I went ahead and merged the latest changes in here to trigger a rebuild of the tests. I'm very :suspect:  at that test failure. I know why it's failing but not *why* it's failing.

```
        cache_file = tempfile.NamedTemporaryFile()
        # Pickles the metadata and stores it in cache_file
>       with salt.utils.files.fopen(str(cache_file), "wb") as fp_:
```

```
>       f_handle = open(*args, **kwargs)  # pylint: disable=resource-leakage
E       OSError: [Errno 22] Invalid argument: '<tempfile._TemporaryFileWrapper object at 0x00000191954B0E08>'
```

Oh. uh... actually no, I know exactly why it's failing, oops. :facepalm: 

Can you make the change to:

```
with salt.utils.files.fopen(cache_file.name, "wb") as fp_:
    ...
```

(in fact - every place it should just be cache_file.name. It's possible that was a holdover from earlier Python, but...)

It's clear this test was not running consistently, or ever(?) because it would have failed every single time, as `str(cache_file)` just provides one of those fairly useless reprs:

```
>>> import tempfile
>>> t = tempfile.NamedTemporaryFile()
>>> str(t)
'<tempfile._TemporaryFileWrapper object at 0x7f88d6dab1c0>'
```

I glanced through the PR otherwise and didn't see any glaring issues, I'll put this PR on my list of reviews for tomorrow :+1:

**Comment by Nerigal** [comment]
ok what the f... again with those tests
im seriously getting tired of it

**Comment by Nerigal** [comment]
most of the test are fixed by the revert i made, but windows still have a permission error that i can't fix

File: requirements/static/ci/common.in [hunk 1]
```diff
@@ -1,7 +1,24 @@
 # Requirements in this file apply to all platforms.
 # We can also exclude platforms from the requirements using markers, but if a requirement only applies
 # to a particular platform, please add it to the corresponding `<platform>.in` file in this directory.
-azure==4.0.0; sys_platform != 'win32'
+#azure==5.0.0; sys_platform != 'win32'
```
Comment (s0undt3ch) [inline requirements/static/ci/common.in:4]: Why is this one commented out and the rest of the direct dependencies added in this file?

File: salt/cloud/clouds/azurearm.py [hunk 2]
```diff
@@ -1296,7 +1342,7 @@ def _query_node_data(name, bootstrap_interface):
         return ip_address
 
     try:
-        data = salt.utils.cloud.wait_for_ip(
+        data = __utils__["cloud.wait_for_ip"](
```
Comment (s0undt3ch) [inline salt/cloud/clouds/azurearm.py:1345]: Why are you using `__utils__` instead of how it was?

File: salt/cloud/clouds/msazure.py [hunk 3]
```diff
@@ -36,7 +36,7 @@
       management_host: management.core.windows.net
 """
 # pylint: disable=function-redefined
-
+# pylint: disable=E1123
```
Comment (s0undt3ch) [inline salt/cloud/clouds/msazure.py:39]: Either remove this disable or the other 3 further down on the try/except?

File: salt/modules/azurearm_compute.py [hunk 4]
```diff
@@ -90,38 +99,49 @@ def availability_set_create_or_update(
...
 
-    compconn = __utils__["azurearm.get_client"]("compute", **kwargs)
+    compconn = salt.utils.azurearm.get_client("compute", **kwargs)
 
     # Use VM names to link to the IDs of existing VMs.
     if isinstance(kwargs.get("virtual_machines"), list):
         vm_list = []
         for vm_name in kwargs.get("virtual_machines"):
-            vm_instance = __salt__["azurearm_compute.virtual_machine_get"](
+            vm_instance = salt.azurearm_compute.virtual_machine_get(
```
Comment (s0undt3ch) [inline salt/modules/azurearm_compute.py:136]: This is not the right import right?

File: salt/utils/azurearm.py [hunk 5]
```diff
@@ -330,3 +332,14 @@ def compare_list_of_dicts(old, new, convert_id_to_name=None):
...
 
     return ret
+
+
+def get_config_from_cloud(cloud_provider):
+    """
+    Function use to retreive the configuration from the cloud
+    provider
+    """
+    conn_kwargs = ""
```
Comment (s0undt3ch) [inline salt/utils/azurearm.py:342]: This variable here is unnecessary

File: tests/unit/pillar/test_azureblob.py [hunk 6]
```diff
@@ -26,6 +27,24 @@
 except ImportError:
     pass
 
+is_windows = sys.platform.startswith("win")
```
Comment (s0undt3ch) [inline tests/unit/pillar/test_azureblob.py:30]: Global variables should be uppercased

**Review by s0undt3ch — CHANGES_REQUESTED** [review]
Correct me if I'm wrong but these changes will require azure >= 5.0.0 right?
Older versions won't work with salt anymore, correct?

File: salt/cloud/clouds/azurearm.py (hunk 2 above)
Comment (Nerigal) [reply salt/cloud/clouds/azurearm.py:1345]: salt.utils. causeing issue with the loader

**Comment by waynew** [comment]
That might be correct... though it's a bit frangible currently as it is. I started looking into the existing tests for this code and they need some help. I'm working with @Nerigal to get this into a bit better state -- we'll have to make sure that we're not needlessly breaking backwards compatibility.

**Comment by Nerigal** [comment]
According to Microsoft, Azure package 5.0.0 and below are deprecated and so i don't see the point of supporting deprecated packages. 
This PR use the most recent and supported upfront api packages that are now provided independently  
Remove all old version of azure arm using pip3 and reinstall the new supported packages 
you can use `pip3 install -r requirement.txt `

**Comment by waynew** [comment]
That's reasonable - we drop support (implicitly or explicitly) for things that are no longer supported, so I'm :+1: for ignoring <5.0

**Comment by waynew** [comment]
Adding help-wanted to this, Nerigal is no longer interested in completing this - if anyone else would like to pick it up and move it across the line, that would be :+1:

**Comment by nicholasmhughes** [comment]
this work is going to move over to https://github.com/salt-extensions/saltext-azurerm

we're taking on the library updates, and we'll also get some of the optimizations in that you've outlined here.

thank you for the effort! closing PR at this time, and stay tuned in the repo above for updates

**Pull request closed without merging**